File: 4.8-bc/ts/stage-2-feature-first/src/features/catalog/repository.ts
import type { DbPool } from "../../shared/db.js";
import type { Product } from "./model.js";

export class Repository {
  constructor(private readonly db: DbPool) {}

  async list(): Promise<Product[]> {
    const res = await this.db.query<{
      id: string;
      name: string;
      price_cents: string;
      category_id: string;
    }>(
      `SELECT id, name, price_cents, category_id FROM catalog_products ORDER BY name`,
    );
    return res.rows.map((r) => ({
      id: r.id,
      name: r.name,
      priceCents: Number(r.price_cents),
      categoryId: r.category_id,
    }));
  }
}

File: 4.8-bc/go/stage-2-feature-first/app/catalog/repository.go
package catalog

import (
	"context"

	"github.com/jackc/pgx/v5/pgxpool"
)

type Repository struct {
	db *pgxpool.Pool
}

func NewRepository(db *pgxpool.Pool) *Repository {
	return &Repository{db: db}
}

func (r *Repository) List(ctx context.Context) ([]Product, error) {
	rows, err := r.db.Query(ctx,
		`SELECT id, name, price_cents, category_id FROM catalog_products ORDER BY name`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []Product
	for rows.Next() {
		var p Product
		if err := rows.Scan(&p.ID, &p.Name, &p.PriceCents, &p.CategoryID); err != nil {
			return nil, err
		}
		out = append(out, p)
	}
	return out, rows.Err()
}
